src/persist/src/indexed/mod.rs

```diff
@@ -198,13 +200,20 @@ impl<K: Data, V: Data, U: Buffer, L: Blob> Indexed<K, V, U, L> {
                 entry
                     .updates
                     .iter()
-                    .map(|((key, val), ts, diff)| (seqno, (key.clone(), val.clone()), *ts, *diff)),
+                    .map(|((key, val), ts, diff)| ((key.clone(), val.clone()), *ts, *diff)),
             );
+
+            // Record the sequence number so we can sanity check that it falls inside the
+            // declared range.
+            sequence_numbers.insert(seqno);
             Ok(())
         })?;
+
+        Self::validate_buf_snapshot(&desc, &sequence_numbers)?;
```

Comment:
Hmm this is still pretty awkward, especially with constructing the unnecessary BTreeSet (unnecessary now that we're not doing the uniqueness check). Would it work to instead pull the following bit all into a method and put the sanity check there?

I do also have a preference for keeping this check debug-only but it's a mild preference

```
            // Future batches are required to be sorted by (ts, k, v).
            updates.sort_unstable_by(|((k1, v1), t1, _), ((k2, v2), t2, _)| {
                (t1, k1, v1).cmp(&(t2, k2, v2))
            });
            let batch = BlobFutureBatch {
                id,
                desc: Description::new(
                    Antichain::from_elem(desc.start),
                    Antichain::from_elem(desc.end),
                    // We never compact BlobFuture, so since is always the minimum.
                    Antichain::from_elem(SeqNo(0)),
                ),
                updates: updates,
            };
            let key = self.new_blob_key();
            self.append_future(key, batch)?;
```

Reply:
I'm happy to do it this way -- the reason I didn't previously is because afaict this approach requires you to make two copies of updates -- one that has the sequence number inlined for the sanity check and then one that doesn't so that you can put it in the future batch. thinking about it more now, I don't think the imagined perf benefits of the btreeset approach justify the clunky interface at this stage, so I'll go ahead and change it

Reply:
Oh I see this issue! Sorry I didn't grok that at first.

Soooo, I've been meaning to make Buffer::snapshot take in a lower bound on SeqNo and move the filtering there, anyway. If we did that, we could just remove the SeqNo argument entirely from the closure and leave everything up to Buffer (which kinda feels more appropriate anyway to me tbh). Concretely, new sig for Buffer::snapshot

```
    /// Returns a consistent snapshot of all written but not yet truncated
    /// entries.
    ///
    /// Takes a closed lower bound on the SeqNos to include and returns an open upper bound on the ones it did include.
    fn snapshot<F>(&self, seqno_lower: SeqNo, logic: F) -> Result<SeqNo, Error>
    where
        F: FnMut(&[u8]) -> Result<(), Error>;
```

Thoughts?

Reply:
hahaha I think we just had a irl race condition because I just pushed a commit that does the change you described.

I'm not sure about removing the sequence number from the closure partially because I'm not sure what guarantees Buffer whats to provide about sequence numbers -- like in that world it seems like sequence numbers are just an opaque generation counter and there's no way for a consumer of snapshot to ask "well did I see _all_ the sequence numbers between [lo, hi)" or just some of them. My instinct is that the interface you've described makes sense and would work, I just want to give my instinct a few minutes to settle down because its been wrong before :D

Reply:
I think we'll make the snapshot change eventually but I'm happy punting for now